Fixing API Spec

jgraeger opened the pull request:
This PR introduces final breaking changes to the specifications and brings the API into a stable state. After that, more changes might be introduced, but not in a breaking way.

The following changes are gonna be affected by changes (checked means ready for review)

- [x] Authencation service 
- [x] Mood diary
- [x] Wiki Service
- [x] User Service
- [x] Motivator Service (Starkmacher)   

For those services, after merging the PR, @kopfsachen-dev/backend shall commit to not introduce any additional services in a breaking way. 

⚠️ Remark:
After the PR got approved, I will mark all endpoints that are not yet in a stable state in any environment as `deprecated`. The deprecation notice gets removed, as soon as a service is deployed in a stable state. After removing deprecation notices from all services/endpoints, the deprecated tag will be used in the intended way.

theEpsilon commented:
>   @theEpsilon: The authentication documentation is updated and I consider it finished. Please review that part.

A few things are missing:

- The request body property `"csrf_token": "<token>"` is required in the browser flow submissions for registration and login
- Each browser-related submit request (Reg, Login, Logout) needs the headers: `Accept: application/json, Content-Type: application/json`
- Login submit flow is missing the request body definition:
`{
	"identifier":"<accountKey>",
	"method": "password",
	"password":"<md5(accountKey)>",
	"csrf_token": "<token>"
	}`

Apart from that, auth part looks good

isaschm commented on schemas/user.yaml:
```diff
@@ -1,65 +1,85 @@
+    - admin
+    - user
+  type: string
+  description: An enumeration.
```
```suggestion
  description: An enumeration of roles.
```
🤔

jgraeger commented:
> > @theEpsilon: The authentication documentation is updated and I consider it finished. Please review that part.
> 
> A few things are missing:

Thanks you very much @theEpsilon. All of your feedback should be implemented. Could you check again? :)

theEpsilon commented:
Might be worth mentioning the JSON headers in every request. If they are not present, the server will send redirects which are undiserable in our use case.

Specifically those headers are required:

`Accept: application/json` in:
- GET /self-service/registration/browser
- GET /self-service/login/browser
- GET /self-service/logout/browser

`Content-Type: application/json` in:
- POST /self-service/registration
- POST /self-service/login

Everything else is fine!

jgraeger commented:
> Cuz it seems like these header declarations are not visible in the spec visualization (?) I can see them in the code though. Everything else is fine!

In OpenAPI v3 you set the a content-type for each operations requestBody and responses. The requester has to set the HTTP `Accepts` and `Content-Type` headers accordingly. Standard compliant code generations tools should also act that way. Therefore, the `x-accepts` and `x-contentType` properties are actually not needed, i've just added them for compatibility with tools not complying fully to the open api spec. Anyway, they are not shown in the Swagger UI specifically as a requirement.

I agree that this can be misleading and added a few more words on the descriptions of the endpoints you mentioned and hope that it's clear now. Thank you for all the work you put into it :)

jgraeger commented:
As I included the requested changes, this PR closes #20.

jgraeger commented:
I consider the spec done and stable for `1.0` release. Please review.

jgraeger commented:
Some remarks on the (semantic) versioning of the API. A major release (First of the two numbers in the version string) indicates the introduction of breaking changes, why a minor release shall be compatible to all prior versions of the API of the same major relase. Thats why i bumped the version to 1.0.

ldb commented on openapi.yaml:
```diff
@@ -1,79 +1,84 @@
-  /wiki/{wikiEntryId}/admin:
-    $ref: 'specs/wiki.yaml#/adminDelete'
+    $ref: 'specs/wiki.yaml#/createEntry'
+  /wiki/{articleId}/admin:
```
```suggestion
  /admin/wiki/{articleId}:
```

We do prefix matching for the AccessRules in Oathkeeper, so in order to prevent access, `admin` should be the first path component.

ldb commented on openapi.yaml:
```diff
@@ -1,79 +1,84 @@
+    $ref: 'specs/wiki.yaml#/getMany'
+  /wiki/{articleId}:
     $ref: 'specs/wiki.yaml#/getOne'
   /wiki/admin:
```
```suggestion
  /admin/wiki:
```

jgraeger replied:
Fixed.

ldb commented on schemas/diary.yaml:
```diff
@@ -1,19 +1,72 @@
-    type: postive
-    description: Today I worked 14h straight for the project. Doing fantastic is literally inevitable
-    timestamp: 2022-06-25T20:03:07+00:00
+    error: mood_day must be a valid ISO 8601 timestamp
```
Does It really?

jgraeger replied:
Yeah. But could also be [RFC2822](https://tools.ietf.org/html/rfc2822#section-3.3). Gonna change that :)

ldb replied:
Why not something more sane like [RFC3339](https://datatracker.ietf.org/doc/html/rfc3339)?
Ultimately I am fine with either, I just don't like ISO8601 at all :D

jgraeger replied:
Because all stakeholders use JavaScript and ISO is the native representation for JS/ES `Date`. It's not that I wouldn't share you're opinion, but in that case, we should let it *Go*.

ldb commented on openapi.yaml:
```diff
@@ -1,79 +1,84 @@
   - name: wiki-service
-  - name: administration-service
+    description: Static wiki articles
   - name: tilt-service
```
Whats this?

isaschm commented on specs/userService.yaml:
```diff
@@ -1,128 +1,71 @@
+              $ref: '../schemas/user.yaml#/UserDataResponse'
       "401":
         description: Unauthorized
+      "422":
```
```suggestion
      "404":
        description: "User not found"
      "422":
```

isaschm reviewed (approved):
I checked the user service schema and specs 👌

MHajoha commented on specs/userService.yaml:
```diff
@@ -144,34 +87,62 @@ userSettingsEndpoint:
+
+      The endpoint is only accessible by administrators and considered **unstable**.
+    operationId: user:admin:getAll
+    deprecated: true
```
Why the deprecation?

jgraeger replied:
> The endpoint is only accessible by administrators and considered **unstable**.

Those endpoints gonna change as we don't have yet a final concept on access control for administrators (offer no scientific research opportunity for our current focus. And furthermore you are not supposed to use it anyway :)

MHajoha reviewed (approved):
Looks good for the most part, thanks. I'm sure once we implement each feature, some questions or necessary changes will still come up, but this looks like a great baseline :+1:

jgraeger commented on schemas/wiki.yaml:
```diff
@@ -1,59 +1,127 @@
     type: string
+    minLength: 24
+    maxLength: 24
   style: form
```
True. Well spotted.

jgraeger commented on openapi.yaml:
```diff
@@ -1,79 +1,84 @@
   - name: user-service
-  - name: notification-service
+    description: Service for storing user data and settings
+  # - name: notification-service
+  #   description: Notifier is used to set up a variety of notifications, e.g. to remind a user of completing an motivator exercise
```
We gonna add a notification service soonish, as it's an interesting one for our scientific focus. But this what you have VCS for. Gonna remove that.

jgraeger commented:
@MHajoha Many thanks for you quick review of the changes and the approval. I've implemented you're feedback :)

I request a frontend review again for the browser teams. Particularly interesting for you could be, that the browser authentication flow is now described in detail in the spec. In case of struggling with the actual implementation, you can get some inspiration from mindtastic/stagefright (credits to @theEpsilon who implemented the API connection on this demo).

Siar-Akbayin commented:
lgtm
